### PR TITLE
Make nll_loss honor ignore_index

### LINKED ISSUE #859: CrossEntropyLoss with ignore_index is not working as expected
Here's an example to replicate the issue: 

```
import torch
import torch_xla_py.xla_model as xm
loss = torch.nn.CrossEntropyLoss(ignore_index=-1)
labels = torch.LongTensor([-1, -1, 3])
preds = torch.ones(3, 5)
print(loss(preds, labels)) # tensor(1.6094)

xla_device = xm.xla_device()
labels = labels.to(xla_device)
preds = preds.to(xla_device)
print(loss(preds, labels))  # expected: tensor(1.6094, device='xla:1')
                                           # printed: tensor(0.5365, device='xla:1')
```

The loss of each individual item is computed correctly (taking the `ignore_index` flag into account), but then the averaging ignores it

### PR DESCRIPTION
**Pull request by asuhan**
Fixes #859.

### CODE REVIEW

File: torch_xla/csrc/nll_loss.cpp [hunk 1]
```diff
@@ -41,49 +41,66 @@ xla::XlaOp LabelsToOneHot(xla::XlaBuilder* builder, xla::int64 depth, int axis,
...
   // Compute sum(-one_hot_labels * logits) / batch.
   xla::XlaOp mul = xla::Mul(xla::Neg(one_hot_labels), logits);
+  xla::XlaOp batch = xla::ConvertElementType(
+      ValidLabelsCount(labels, ignore_index), logits_element_type);
   xla::XlaComputation add_func =
-      XlaHelpers::CreateAddComputation(logits_shape.element_type());
-  xla::XlaOp batch = XlaHelpers::ScalarValue<float>(
-      logits_shape.dimensions(0), logits_shape.element_type(), builder);
+      XlaHelpers::CreateAddComputation(logits_element_type);
   return xla::ReduceAll(mul, zero, add_func) / batch;
```
Comment (dlibenzi) [inline torch_xla/csrc/nll_loss.cpp]: I may be mistaken, but I think the values with ignore_index should not be reduced.
IOW, doing the proper "batch" division is only one part of the job.
Also, how is our one-hot WRT negative class indices?
Can you try to test with an ignore_index within the range of the class indices (say, ignore_index == 5 in a 10 class setup)?

File: test/cpp/test_aten_xla_tensor.cpp [hunk 2]
```diff
@@ -5365,23 +5365,25 @@ TEST_F(AtenXlaTensorTest, TestConv2DNonSquare) {
 }
 
 TEST_F(AtenXlaTensorTest, TestNllLoss) {
-  int batch = 3;
+  int batch = 20;
```
[the remark this answers is not in the document]
Comment (asuhan) [reply test/cpp/test_aten_xla_tensor.cpp]: If we use a smaller batch size, we might get unlucky and not get any indices to be ignored. Other ways aren't as straightforward as I'd like, what would be your suggestion here?

File: torch_xla/csrc/nll_loss.cpp (hunk 1 above)
Comment (asuhan) [reply torch_xla/csrc/nll_loss.cpp]: I'll double check this. If it's outside the range, it'll work out ok since one-hot will be all zeros for that batch, but if it's inside the range you might have be right.

File: test/cpp/test_aten_xla_tensor.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (dlibenzi) [reply test/cpp/test_aten_xla_tensor.cpp]: Can we limit the range of randint()?

[the remark this answers is not in the document]
Comment (asuhan) [reply test/cpp/test_aten_xla_tensor.cpp]: Done.

File: torch_xla/csrc/nll_loss.cpp [hunk 3]
```diff
@@ -41,49 +41,66 @@ xla::XlaOp LabelsToOneHot(xla::XlaBuilder* builder, xla::int64 depth, int axis,
...
+
 }  // namespace
 
 // Builds the NLLLoss for log-probabilities "logits" and class indices "labels".
-xla::XlaOp BuildNllLoss(const xla::XlaOp& logits, const xla::XlaOp& labels) {
+xla::XlaOp BuildNllLoss(const xla::XlaOp& logits, const xla::XlaOp& labels,
+                        int ignore_index) {
   xla::XlaBuilder* builder = logits.builder();
   xla::Shape logits_shape = XlaHelpers::ShapeOfXlaOp(logits);
+  xla::PrimitiveType logits_element_type = logits_shape.element_type();
```
[the remark this answers is not in the document]
Comment (asuhan) [reply torch_xla/csrc/nll_loss.cpp]: Done.

File: torch_xla/csrc/nll_loss.cpp [hunk 4]
```diff
@@ -41,49 +41,66 @@ xla::XlaOp LabelsToOneHot(xla::XlaBuilder* builder, xla::int64 depth, int axis,
                      xla::Broadcast(off_value, output_dimensions));
 }
 
+// Count the number of labels which aren't equal to "ignore_index".
+xla::XlaOp ValidLabelsCount(const xla::XlaOp& labels, int ignore_index) {
+  xla::XlaBuilder* builder = labels.builder();
```
[the remark this answers is not in the document]
Comment (asuhan) [reply torch_xla/csrc/nll_loss.cpp]: Done.

File: torch_xla/csrc/nll_loss.cpp (hunk 1 above)
Comment (asuhan) [reply torch_xla/csrc/nll_loss.cpp]: Done, added a path for the case when `ignore_index` is a valid class index.

**Comment by asuhan** [comment]
I couldn't think of an easier way, alternatives (scatter) were more complicated and the tensor we materialize should be tiny.
